File: 235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(var _value: Int) {
 *   var value: Int = _value
 *   var left: TreeNode = null
 *   var right: TreeNode = null
 * }
 */

object Solution {
    def lowestCommonAncestor(root: TreeNode, p: TreeNode, q: TreeNode): TreeNode = {
        if (root == p || root == q || (root.value > q.value && root.value < p.value) || (root.value > p.value && root.value < q.value)) root
        else if (root.value < p.value) lowestCommonAncestor(root.right, p, q)
        else lowestCommonAncestor(root.left, p, q)
    }
}

File: 235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        TreeNode result = root;
        
        while(root != null) {
            if (p.val == root.val || q.val == root.val) {
                return root;
            }
            
            if (p.val < root.val && q.val < root.val) {
                result = root;
                root = root.left;
            } else if (p.val > root.val && q.val > root.val) {
                result = root;
                root = root.right;
            } else {
                return root;
            }
        }
        
        return result;
    }
}